File: db-4.8.30/examples_cxx/EnvExample.cpp
/*-
 * See the file LICENSE for redistribution information.
 *
 * Copyright (c) 1997-2009 Oracle.  All rights reserved.
 *
 * $Id$
 */

#include <sys/types.h>

#include <errno.h>
#include <iostream>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <db_cxx.h>

using std::ostream;
using std::cout;
using std::cerr;

void db_setup(const char *, const char *, ostream&);
void db_teardown(const char *, const char *, ostream&);
static int usage();

const char *progname = "EnvExample";            /* Program name. */

//
// An example of a program creating/configuring a Berkeley DB environment.
//
int
main(int argc, char *argv[])
{
    //
    // Note: it may be easiest to put all Berkeley DB operations in a
    // try block, as seen here.  Alternatively, you can change the
    // ErrorModel in the DbEnv so that exceptions are never thrown
    // and check error returns from all methods.
    //
    try {
        const char *data_dir, *home;
        
        //
        // All of the shared database files live in home,
        // but data files live in data_dir.
        //
        home = "TESTDIR";
        data_dir = "data";

        for (int argnum = 1; argnum < argc; ++argnum) {
            if (strcmp(argv[argnum], "-h") == 0) {
                if (++argnum >= argc)
                    return (usage());
                home = argv[argnum];
            }
            else if (strcmp(argv[argnum], "-d") == 0) {
                if (++argnum >= argc)
                    return (usage());
                data_dir = argv[argnum];
            }
            else {
                return (usage());
            }
        }

        cout << "Setup env\n";
        db_setup(home, data_dir, cerr);

        cout << "Teardown env\n";
        db_teardown(home, data_dir, cerr);
        return (EXIT_SUCCESS);
    }
    catch (DbException &dbe) {
        cerr << "EnvExample: " << dbe.what() << "\n";
        return (EXIT_FAILURE);
    }
}

// Note that any of the db calls can throw DbException
void
db_setup(const char *home, const char *data_dir, ostream& err_stream)
{
    //
    // Create an environment object and initialize it for error
    // reporting.
    //
    DbEnv *dbenv = new DbEnv(0);
    dbenv->set_error_stream(&err_stream);
    dbenv->set_errpfx(progname);

    //
    // We want to specify the shared memory buffer pool cachesize,
    // but everything else is the default.
    //
    dbenv->set_cachesize(0, 64 * 1024, 0);

    // Databases are in a subdirectory.
    (void)dbenv->set_data_dir(data_dir);

    // Open the environment with full transactional support.
    dbenv->open(home,
        DB_CREATE | DB_INIT_LOCK | DB_INIT_LOG | DB_INIT_MPOOL | 
        DB_INIT_TXN, 0);

    // Open a database in the environment to verify the data_dir
    // has been set correctly.
    // Create a database handle, using the environment. 
    Db *db = new Db(dbenv, 0) ;

    // Open the database. 
    db->open(NULL, "EvnExample_db1.db", NULL, DB_BTREE, DB_CREATE, 0644);

    // Close the database handle.
    db->close(0) ;
    delete db;

    // Close the handle.
    dbenv->close(0);
    delete dbenv;
}

void
db_teardown(const char *home, const char *data_dir, ostream& err_stream)
{
    // Remove the shared database regions.
    DbEnv *dbenv = new DbEnv(0);

    dbenv->set_error_stream(&err_stream);
    dbenv->set_errpfx(progname);

    (void)dbenv->set_data_dir(data_dir);
    dbenv->remove(home, 0);
    delete dbenv;
}

static int
usage()
{
    cerr << "usage: excxx_env [-h home] [-d data_dir]\n";
    return (EXIT_FAILURE);
}

File: db-4.8.30/examples_java/src/db/EnvExample.java
/*-
 * See the file LICENSE for redistribution information.
 *
 * Copyright (c) 1997-2009 Oracle.  All rights reserved.
 *
 * $Id$
 */

package db;

import com.sleepycat.db.*;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.OutputStream;

/*
 * An example of a program configuring a database environment.
 *
 * For comparison purposes, this example uses a similar structure
 * as examples/ex_env.c and examples_cxx/EnvExample.cpp.
 */
public class EnvExample {
    private static final String progname = "EnvExample";

    private static void runApplication(Environment dbenv)
        throws DatabaseException, FileNotFoundException {
         
        // Open a database in the environment to verify the data_dir
        // has been set correctly.
        DatabaseConfig dbconfig = new DatabaseConfig();

        // The database is DB_BTREE.
        dbconfig.setAllowCreate(true);
        dbconfig.setMode(0644);
        dbconfig.setType(DatabaseType.BTREE);
        Database db=dbenv.openDatabase(null, 
            "jEnvExample_db1.db", null, dbconfig);

        // Close the database.
        db.close();
    }

    private static void setupEnvironment(File home,
                                         File dataDir,
                                         OutputStream errs)
        throws DatabaseException, FileNotFoundException {

        // Create an environment object and initialize it for error reporting.
        EnvironmentConfig config = new EnvironmentConfig();
        config.setErrorStream(errs);
        config.setErrorPrefix(progname);

        //
        // We want to specify the shared memory buffer pool cachesize,
        // but everything else is the default.
        //
        config.setCacheSize(64 * 1024);

        // Databases are in a separate directory.
        config.addDataDir(dataDir);

        // Open the environment with full transactional support.
        config.setAllowCreate(true);
        config.setInitializeCache(true);
        config.setTransactional(true);
        config.setInitializeLocking(true);

        //
        // open is declared to throw a FileNotFoundException, which normally
        // shouldn't occur when allowCreate is set.
        //
        Environment dbenv = new Environment(home, config);

        try {
            // Start your application.
            runApplication(dbenv);
        } finally {
            // Close the environment.  Doing this in the finally block ensures
            // it is done, even if an error is thrown.
            dbenv.close();
        }
    }

    private static void teardownEnvironment(File home,
                                            File dataDir,
                                            OutputStream errs)
        throws DatabaseException, FileNotFoundException {

        // Remove the shared database regions.
        EnvironmentConfig config = new EnvironmentConfig();

        config.setErrorStream(errs);
        config.setErrorPrefix(progname);
        config.addDataDir(dataDir);
        Environment.remove(home, true, config);
    }

    private static void usage() {
        System.err.println("usage: java db.EnvExample [-h home] [-d data_dir]");
        System.exit(1);
    }

    public static void main(String[] argv) {
        //
        // All of the shared database files live in home,
        // but data files live in dataDir.
        //
        // Using Berkeley DB in C/C++, we need to allocate two elements
        // in the array and set config[1] to NULL.  This is not
        // necessary in Java.
        //
        File home = new File("TESTDIR");
        File dataDir = new File("data");

        for (int i = 0; i < argv.length; ++i) {
            if (argv[i].equals("-h")) {
                if (++i >= argv.length)
                    usage();
                home = new File(argv[i]);
            } else if (argv[i].equals("-d")) {
                if (++i >= argv.length)
                    usage();
                dataDir = new File(argv[i]);
            } else if (argv[i].equals("-u")) {
                usage();
            }
        }

        try {
            System.out.println("Setup env");
            setupEnvironment(home, dataDir, System.err);

            System.out.println("Teardown env");
            teardownEnvironment(home, dataDir, System.err);
        } catch (DatabaseException dbe) {
            System.err.println(progname + ": environment open: " + dbe.toString());
            System.exit (1);
        } catch (FileNotFoundException fnfe) {
            System.err.println(progname + ": unexpected open environment error  " + fnfe);
            System.exit (1);
        }
    }

}
